DS-195 remove the use of keystore apis on add a validator doc

Julian-dev28 opened the pull request:
- Replaces `Add a validator with API calls` with `Add a validator with AvalancheJS` 
- Fixes a spelling error from a previous push

yulin-dong reviewed (changes requested):
static/img/ajs-addValidator.png is not used. please remove

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
````diff
@@ -74,156 +75,332 @@ Once your validator’s start time has passed, you will see the rewards it may e
+or add it to an existing project:
+
+```zsh
+$ yarn add --dev avalanche
````
why `--dev` here?

Julian-dev28 replied:
I saw this workflow from a non-production point of view so it seems more accurate.

Open to changing

yulin-dong replied:
https://classic.yarnpkg.com/lang/en/docs/cli/add/
`Using --dev or -D will install one or more packages in your [devDependencies](https://classic.yarnpkg.com/en/docs/dependency-types#toc-dev-dependencies).`

Are you suggesting developers to just use avalancheJS in devdependencies?

jpop32 commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -74,156 +75,332 @@ Once your validator’s start time has passed, you will see the rewards it may e
+
+An array of addresses as a [Buffer](https://github.com/feross/buffer) who own the staking UTXOs.
+
+`changeAddresses`
```
Are `change`, `to` and `reward` really arrays? On the platform.addValidator() those are all single address, not an array of them.

Julian-dev28 replied:
This was pulled from `avalanchejs/avalanche/dist/apis/platformvm/api.d.ts` 

```js
     * @param toAddresses An array of addresses as {@link https://github.com/feross/buffer|Buffer} who received the staked tokens at the end of the staking period
     * @param fromAddresses An array of addresses as {@link https://github.com/feross/buffer|Buffer} who own the staking UTXOs the fees in AVAX
     * @param changeAddresses An array of addresses as {@link https://github.com/feross/buffer|Buffer} who gets the change leftover from the fee payment
     * @param rewardAddresses The addresses which will receive the rewards from the delegated stake.
```

Also, I tried to pass a single string as an argument [here](https://github.com/ava-labs/avalanchejs/blob/ffaf4d2ef2bbc62a57685d611e3a98115a2d2195/examples/platformvm/buildAddValidatorTx.ts#L41-L50) but the compiler asked for a string[]

Could use clarity on this

jpop32 commented on docs/nodes/validate/add-a-validator.md:
````diff
@@ -74,156 +75,332 @@ Once your validator’s start time has passed, you will see the rewards it may e
+const privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`;
+```
+
+and replace this with a private key that you control.
````
Maybe link to appropriate example for getting a pirvate key? https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/createKeypair.ts

yulin-dong replied:
Please update based on the comment above.

Julian-dev28 replied:
committed

https://github.com/ava-labs/avalanche-docs/blob/a8a9c63fe3e51b34e1bcfb0b8c2be41c3bcbf13b/docs/nodes/validate/add-a-validator.md?plain=1#L103

jpop32 reviewed (changes requested):
A few nits.

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
````diff
@@ -74,156 +75,332 @@ Once your validator’s start time has passed, you will see the rewards it may e
+
+The sample code assigns the variable, `stakeAmount` to be the minimum amount of AVAX that a node must stake to become a validator(1 AVAX).
+```ts
+const main = async (): Promise<any> => {
+  const stakeAmount: any = await pchain.getMinStake()
+  ...
+}
+```
+
+To stake a different amount, please replace the code above with a stake amount of 1 AVAX or greater. Below sets a new value of 10,000 AVAX (`10000000000000` GWEI). Value is set in GWEI format where `1,000,000,000` GWEI = 1 AVAX. 
+
+Please follow the Example below for steps to change the stake amount:
+
+Locate this line in the file
+
+```js
+  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
+    ...
+    stakeAmount.minValidatorStake,
+    ...
+  )
+```
+
+and change the default value to a new `BN` value, for example:
+
+```js
+  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
+    ...
+    new BN(1e12),
````
This is a convoluted way to say that they just need to put a new value for the stake amount as `BN(1e12)`

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
````diff
@@ -74,156 +75,332 @@ Once your validator’s start time has passed, you will see the rewards it may e
+```
+
+
+#### Start and End Time
````
This is a duplicate section above (Line 200-203)

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
````diff
@@ -48,182 +48,308 @@ The response has your node’s ID:
+
+Avalanche allows for delegation of stake. This parameter is the percent fee this validator charges when others delegate stake to them. For example, if `delegationFeeRate` is `10` and someone delegates to this validator, then when the delegation period is over, 10% of the reward goes to the validator and the rest goes to the delegator.
+
+#### Stake Amount
+
+In order to validate the Fuji Network, one must stake AVAX.
+
+The sample code assigns the variable, `stakeAmount` to be the minimum amount of AVAX that a node must stake to become a validator(1 AVAX).
+
+```ts
+const stakeAmount: any = await pchain.getMinStake();
+```
+
+To stake a different amount, please replace the code above with a stake amount of 1 AVAX or greater. Below sets a new value of 10,000 AVAX (`10000000000000` GWEI). Value is set in GWEI format where `1,000,000,000` GWEI = 1 AVAX.
+
+Please follow the Example below for steps to change the stake amount:
+
+Locate this line in the file
+
+```js
+  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
+    ...
+    stakeAmount.minValidatorStake,
+    ...
+  )
+```
+
+and change the default value to a new `BN` value, for example:
+
+```js
+  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
+    ...
+    new BN(1e12),
+    ...
+  )
+```
````
This is a convoluted way to say that they just need to put a new value for the stake amount as BN(1e12). Please rewrite this section to make it short to the point that they just need to provide a correct amount to the API.

Julian-dev28 replied:
I have read this multiple times and I personally feel that this is short and clear. I have removed some language but am open to more suggestions.

yulin-dong replied:
This whole section can be reduced to one paragraph/sentence like this:

Set the the proper staking amount when calling pchain.buildAddValidatorTx by replacing `stakeAmount.minValidatorStake` with a number such as `BN(1e12)` which is 10,000 AVAX, in the example code.

patrick-ogrady commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -16,17 +16,17 @@ Note that once you issue the transaction to add a node as a validator, there is
 ## Requirements
 
-You've completed [Run an Avalanche Node](../build/run-avalanche-node-manually.md) and are familiar with [Avalanche's architecture](../../overview/getting-started/avalanche-platform.md). In this tutorial, we use [Avalanche’s Postman collection](https://github.com/ava-labs/avalanche-postman-collection) to help us make API calls.
+You've completed [Run an Avalanche Node](../build/run-avalanche-node-manually.md) and are familiar with [Avalanche's architecture](../../overview/getting-started/avalanche-platform.md).In this tutorial, we use [AvalancheJS](../../apis/avalanchejs/README.md) and [Avalanche’s Postman collection](https://github.com/ava-labs/avalanche-postman-collection) to help us make API calls.
```
nit: you are missing a space after period:
```
md).In this tutorial, we use
```

patrick-ogrady commented on docs/nodes/validate/add-a-validator.md:
````diff
@@ -48,182 +48,308 @@ The response has your node’s ID:
+and replace this with a private key that you control.
+
+```js
+const privKey: string = "<YOUR-PRIVATE-KEY-HERE>";
````
You should add a node about trying to use a Ledger WHENEVER you can.

Julian-dev28 replied:
Definitely!

I was unaware that we can use a ledger to do this. 
Is [this](https://docs.avax.network/apis/avalanche-wallet-sdk/wallet-classes/#ledger-wallet) something I can use to illustrate using a ledger?

yulin-dong replied:
This will be addressed by https://ava-labs.atlassian.net/browse/DS-208

Julian-dev28 commented:
Commit [ae4cc697a0bf47cb07e5e7731a0bccfa5cc67e08](https://github.com/ava-labs/avalanche-docs/pull/827/commits/ae4cc697a0bf47cb07e5e7731a0bccfa5cc67e08) addresses all comments as of 07JUNE2022 2200

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -48,182 +48,306 @@ The response has your node’s ID:
+Open your AvalancheJS directory and select the [**`examples/platformvm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/platformvm) folder to view the source code for the examples scripts.
 
-This method’s signature is:
+We will use the [**`buildAddValidatorTx.ts`**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildAddValidatorTx.ts) script to add a validator.
```
Also please link to the avalancheJS API `pchain.buildAddValidatorTx` which is the focus point of this section.

yulin-dong replied:
@cgcardona what is your suggestion on linking to the API signature?

yulin-dong replied:
We should add a link to the API signature at https://docs.avax.network/apis/avalanchejs/classes/api_platformvm.platformvmapi#buildaddvalidatortx using the correct relative path.

Julian-dev28 replied:
Roger. I am working on that now

Julian-dev28 replied:
@yulin-dong

is this excerpt correct? if so which one is more accurate?

To learn more about the `buildAddValidatorTx` API, please click [here](../../apis/avalanchejs/classes/api_platformvm.platformvmapi.md#buildaddvalidatortx).


To learn more about the API behind `buildAddValidatorTx`, please click [here](../../apis/avalanchejs/classes/api_platformvm.platformvmapi.md#buildaddvalidatortx).

yulin-dong replied:
To learn more about the buildAddValidatorTx API, please click [here](https://github.com/ava-labs/apis/avalanchejs/classes/api_platformvm.platformvmapi.md#buildaddvalidatortx).

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -48,182 +48,306 @@ The response has your node’s ID:
+Open your AvalancheJS directory and select the [**`examples/platformvm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/platformvm) folder to view the source code for the examples scripts.
 
-This method’s signature is:
+We will use the [**`buildAddValidatorTx.ts`**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildAddValidatorTx.ts) script to add a validator.
```
Also please link to the avalancheJS API `pchain.buildAddValidatorTx` which is the focus point of this section.

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -48,182 +48,306 @@ The response has your node’s ID:
+    pAddressStrings,
+    pAddressStrings,
+...
+    pAddressStrings,
```
Which one is for which of the following: `toAddresses`, `fromAddresses`, `changeAddress` and `rewardAddresses`. I am totally lost here.

yulin-dong replied:
Linking to the API signature is critical here so that it is very clear which params is for what.

Julian-dev28 replied:
Definitely. Please refer to the comment above regarding `pchain.buldAddValditor.ts`

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -48,182 +48,306 @@ The response has your node’s ID:
+    ["<YOUR-PCHAIN-ADDRESS-HERE>"],
+...
+    ["<YOUR-PCHAIN-ADDRESS-HERE>"],
+...
```
Same question above, which address is for each of `toAddresses`, `fromAddresses`, `changeAddress` and `rewardAddresses`?

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
````diff
@@ -48,182 +48,306 @@ The response has your node’s ID:
+
+```js
+const main = async (): Promise<any> => {
+ const txID: string = "<YOUR-TXID-HERE>"
````
```suggestion
 const txID: string = "2ftDVwmss5eJk8HFsNVi6a3vWK9s3szZFhEeSY2HCS8xDb8Cra"
```

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -48,182 +48,304 @@ The response has your node’s ID:
+
+`toAddresses`
+
+An array of addresses as [Buffer](https://github.com/feross/buffer) who receive the staked tokens at the end of the staking period.
```
Please provide example of using [Buffer](https://github.com/feross/buffer) if that is the case.

Julian-dev28 replied:
I realize this is confusing. I believe that it would be better to just remove the mention of `Buffer` as it pertains more to how data is parsed and doesnt have an effect on how addresses are passed in.

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -48,182 +48,305 @@ The response has your node’s ID:
+#### Stake Amount
 
-When one issues a transaction to join the Primary Network they specify the time they will enter (start validating) and leave (stop validating.) The minimum duration that one can validate the Primary Network is 2 weeks, and the maximum duration is one year. One can re-enter the Primary Network after leaving, it’s just that the maximum _continuous_ duration is one year. `startTime` and `endTime` ar ...
+Set the the proper staking amount when calling `pchain.buildAddValidatorTx` by replacing `stakeAmount.minValidatorStake` with a number such as `BN(1e12)` which is 10,000 AVAX, in the example code.
```
```suggestion
Set the the proper staking amount in calling `pchain.buildAddValidatorTx` by replacing `stakeAmount.minValidatorStake` with a number in the unit of gwei, for example, `BN(1e12)` which is 10,000 AVAX.
```

Julian-dev28 commented:
commit a5f7c2512a86167f654b544bb4ebb0b9c2dfc8ef adds reference to correct api signature for [buildaddvalidatortx](https://docs.avax.network/apis/avalanchejs/classes/api_platformvm.platformvmapi#buildaddvalidatortx)

Julian-dev28 reviewed (commented):
Changes acknowledged

thank you

cgcardona commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -332,16 +257,92 @@ $ ts-node examples/platformvm/getPendingValidators.ts
+const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
+  utxoSet,
+  [pAddressStrings[0], pAddressStrings[1]], // toAddresses, one or more addresses
+  [], // fromAddresses
```
The `fromAddresses` and `changeAddresses` cannot be empty arrays. They must have at least 1 address. I suggest just passing in an array which contains only the first address from `pAddressStrings`&mdash;`[pAddressStrings[0]]`.

yulin-dong replied:
addressed

cgcardona commented on docs/nodes/validate/add-a-validator.md:
````diff
@@ -332,16 +257,92 @@ $ ts-node examples/platformvm/getPendingValidators.ts
+Locate this part of the code
+
+```ts
+let privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`;
````
In our example code we don't use semicolons so please remove the semicolons from each of these example snippets.

yulin-dong replied:
@Julian-dev28 please update

yulin-dong commented on docs/nodes/validate/add-a-validator.md:
```diff
@@ -48,182 +48,304 @@ The response has your node’s ID:
+  "P-custom18jma8ppw3nhx5r4ap8clazz0dps7rv5u9xde7p", // pAddressStrings[0]
+  "P-custom186e8ee4qnhnazrnyv2k9u44607ux70syngdz6l", // pAddressStrings[1]
+  "P-custom1pdu8d8ss5p329l4vtn23p25cjr57qt5yw2cnsn", // pAddressStrings[2]
+  "P-custom152s2upkznxeucwnp68wmg6wmy863qd2sz6ktww", // pAddressStrings[3]
```
This is not the correct address for Fuji as this section is still under the Fuji Workflow. @Julian-dev28 please change the text how to generate the correct address value for Fuji.

jpop32 reviewed (approved):
Looks good to me now, let's ship it. Further nits can be addressed once this is live.
